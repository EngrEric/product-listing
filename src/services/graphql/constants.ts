import { gql } from "@apollo/client";

export const productTypeDefs = gql`
extend type Product {
    imgUrl: String;
    name: String;
    price: String;
    longDescription: String;
    rating: String;
    brand: String;
    options: String;
    sizes: String;
    id: String;
}
`;
