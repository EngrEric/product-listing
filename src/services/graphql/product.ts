import { gql } from "@apollo/client";

export const productsQuery = {
  GET_ALL_PRODUCTS: gql`
    query products {
      products {
        imgUrl
        name
        price
        shortDescription
        rating
        id
      }
    }
  `,

  GET_ONE_PRODUCT: gql`
    query product($id: ID!) {
      product(id: $id) {
        imgUrl
        name
        price
        longDescription
        rating
        brand
        options
        sizes
        id
      }
    }
  `
};
