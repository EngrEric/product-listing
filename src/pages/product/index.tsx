import { FC } from "react";
import { useNavigate, useParams } from "react-router-dom";
import { useQuery } from "@apollo/client";
import { Button, Stack, Typography } from "@mui/material";

import { ProductDetails } from "../../components/products";
import { GetOneProductsData } from "../../components/products/constants";
import { LoadingSpinner } from "../../components/common";
import { productsQuery } from "../../services/graphql/product";

/**
 * @description The product details page
 */
const ProductDetailsPage: FC = () => {
  const { id: productId } = useParams<string>();
  const navigate = useNavigate();

  const { loading, data, error } = useQuery<GetOneProductsData>(
    productsQuery.GET_ONE_PRODUCT,
    {
      skip: !productId,
      variables: { id: productId }
    }
  );

  if (loading) {
    return (
      <Stack
        direction="column"
        justifyContent="center"
        alignItems="center"
        style={{ minHeight: "60vh" }}
      >
        <LoadingSpinner /> Loading...
      </Stack>
    );
  }
  if (error)
    return (
      <Stack
        direction="column"
        justifyContent="center"
        alignItems="center"
        style={{ minHeight: "60vh" }}
      >
        <Typography
          color={"error"}
          textAlign={"center"}
          pt="100px"
          variant="h5"
          component={"p"}
        >
          Error: {error.message}
        </Typography>
        <Button sx={{ my: 10 }} onClick={() => navigate(-1)}>
          Go back
        </Button>
      </Stack>
    );

  return data ? <ProductDetails product={data.product} /> : <></>;
};

export default ProductDetailsPage;
