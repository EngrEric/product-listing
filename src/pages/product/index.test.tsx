import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";

import ProductDetailsPage from "./";
import { productsQuery } from "../../services/graphql/product";

const mockProductId = "1";
const mockProduct = {
  imgUrl: "mock url",
  name: "Mock product",
  price: "mock price",
  longDescription: "mock long des",
  rating: 4,
  brand: "mock brnad",
  options: '[{"text":"option1","value":"option1"}]',
  sizes: '[{"text":"XL","value":"xl"}]',
  id: "1"
};

describe("ProductDetailsPage", () => {
  const RootComponent = ({ mockType }: { mockType: MockedResponse[] }) => (
    <MockedProvider mocks={mockType} addTypename={false}>
      <MemoryRouter initialEntries={[`/details/${mockProductId}`]}>
        <Routes>
          <Route path="/details/:id" element={<ProductDetailsPage />} />
        </Routes>
      </MemoryRouter>
    </MockedProvider>
  );

  const mocks: MockedResponse[] = [
    {
      request: {
        query: productsQuery.GET_ONE_PRODUCT,
        variables: { id: mockProductId }
      },
      result: {
        data: {
          product: mockProduct
        }
      }
    }
  ];

  it("renders loading spinner while fetching data", async () => {
    render(<RootComponent mockType={[{ ...mocks[0], delay: 30 }]} />);

    expect(screen.getByTestId("loading-spinner")).toBeInTheDocument();
  });

  it('renders error message and "Go back" button on error', async () => {
    const errorMessage = "Mock error message";
    const errorMocks = [
      {
        request: {
          query: productsQuery.GET_ONE_PRODUCT,
          variables: { id: mockProductId }
        },
        error: new Error(errorMessage)
      }
    ];

    render(<RootComponent mockType={errorMocks} />);

    expect(
      await screen.findByText(`Error: ${errorMessage}`)
    ).toBeInTheDocument();

    fireEvent.click(await screen.findByText("Go back"));
    const currentPath = window.location.pathname;
    expect(currentPath).toBe("/");
  });

  it("renders product details when data is loaded", async () => {
    const { asFragment } = render(<RootComponent mockType={mocks} />);

    expect(await screen.findByText(mockProduct.name)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
