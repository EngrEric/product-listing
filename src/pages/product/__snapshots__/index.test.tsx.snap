// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailsPage renders product details when data is loaded 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
    data-testid="product-details"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1sx5805-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ArrowBackIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
       Back
    </button>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-d39eye-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-1l4w6pd"
        >
          <img
            alt="Mock product"
            src="mock url"
            width="100%"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-1ou72n1"
        >
          <h3
            class="MuiTypography-root MuiTypography-h4 css-h5tjps-MuiTypography-root"
          >
            Mock product
          </h3>
          <p
            class="MuiTypography-root MuiTypography-subtitle1 css-i1fd1-MuiTypography-root"
          >
            mock long des
          </p>
          <p
            class="MuiTypography-root MuiTypography-h6 css-9j38tj-MuiTypography-root"
          >
            Brand: 
            <span
              class="MuiTypography-root MuiTypography-body1 css-1mcm3y0-MuiTypography-root"
            >
              mock brnad
            </span>
          </p>
          <div
            class="MuiStack-root css-m69qwo-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-h6 css-2yqopb-MuiTypography-root"
            >
              Rated:
            </p>
            <div
              class="MuiBox-root css-jm3i9k"
              data-testid="rating-component"
            >
              <span
                class="MuiRating-root MuiRating-sizeMedium css-iljtu2-MuiRating-root"
              >
                <span
                  class="MuiRating-decimal css-34he1w-MuiRating-decimal"
                >
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r1:"
                    style="width: 0%; overflow: hidden; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Useless
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":r1:"
                    name="rating-feedback"
                    type="radio"
                    value="0.5"
                  />
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r2:"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Useless+
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":r2:"
                    name="rating-feedback"
                    type="radio"
                    value="1"
                  />
                </span>
                <span
                  class="MuiRating-decimal css-34he1w-MuiRating-decimal"
                >
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r3:"
                    style="width: 0%; overflow: hidden; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Poor
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":r3:"
                    name="rating-feedback"
                    type="radio"
                    value="1.5"
                  />
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r4:"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Poor+
                    </span>
                  </label>
                  <input
                    checked=""
                    class="MuiRating-visuallyHidden"
                    id=":r4:"
                    name="rating-feedback"
                    type="radio"
                    value="2"
                  />
                </span>
                <span
                  class="MuiRating-decimal css-34he1w-MuiRating-decimal"
                >
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r5:"
                    style="width: 0%; overflow: hidden; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        style="opacity: 0.55;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Ok
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":r5:"
                    name="rating-feedback"
                    type="radio"
                    value="2.5"
                  />
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r6:"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        style="opacity: 0.55;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Ok+
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":r6:"
                    name="rating-feedback"
                    type="radio"
                    value="3"
                  />
                </span>
                <span
                  class="MuiRating-decimal css-34he1w-MuiRating-decimal"
                >
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r7:"
                    style="width: 0%; overflow: hidden; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        style="opacity: 0.55;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Good
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":r7:"
                    name="rating-feedback"
                    type="radio"
                    value="3.5"
                  />
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r8:"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        style="opacity: 0.55;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Good+
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":r8:"
                    name="rating-feedback"
                    type="radio"
                    value="4"
                  />
                </span>
                <span
                  class="MuiRating-decimal css-34he1w-MuiRating-decimal"
                >
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":r9:"
                    style="width: 0%; overflow: hidden; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        style="opacity: 0.55;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Excellent
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":r9:"
                    name="rating-feedback"
                    type="radio"
                    value="4.5"
                  />
                  <label
                    class="css-dqr9h-MuiRating-label"
                    for=":ra:"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarIcon"
                        focusable="false"
                        style="opacity: 0.55;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiRating-visuallyHidden"
                    >
                      Excellent+
                    </span>
                  </label>
                  <input
                    class="MuiRating-visuallyHidden"
                    id=":ra:"
                    name="rating-feedback"
                    type="radio"
                    value="5"
                  />
                </span>
                <label
                  class="MuiRating-label MuiRating-pristine css-dqr9h-MuiRating-label"
                >
                  <input
                    class="MuiRating-visuallyHidden"
                    id="rating-feedback-empty"
                    name="rating-feedback"
                    type="radio"
                    value=""
                  />
                  <span
                    class="MuiRating-visuallyHidden"
                  >
                    Empty
                  </span>
                </label>
              </span>
              <div
                class="MuiBox-root css-d0uhtl"
              >
                Poor+
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root css-c9bted"
          >
            <p
              class="MuiTypography-root MuiTypography-h6 css-tjvwag-MuiTypography-root"
            >
              Options
            </p>
            <div
              aria-label="text alignment"
              class="MuiToggleButtonGroup-root css-1vc0b5m-MuiToggleButtonGroup-root"
              role="group"
            >
              <button
                aria-pressed="true"
                class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-14zpu5i-MuiButtonBase-root-MuiToggleButton-root"
                tabindex="0"
                type="button"
                value="option1"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  option1
                </p>
              </button>
            </div>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <p
              class="MuiTypography-root MuiTypography-h6 css-tjvwag-MuiTypography-root"
            >
              Sizes
            </p>
            <div
              aria-label="text alignment"
              class="MuiToggleButtonGroup-root css-1vc0b5m-MuiToggleButtonGroup-root"
              role="group"
            >
              <button
                aria-pressed="true"
                class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-14zpu5i-MuiButtonBase-root-MuiToggleButton-root"
                tabindex="0"
                type="button"
                value="xl"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  XL
                </p>
              </button>
            </div>
          </div>
          <div
            class="MuiBox-root css-c9bted"
          >
            <h3
              class="MuiTypography-root MuiTypography-h5 css-1m0ncb0-MuiTypography-root"
            >
              mock price
            </h3>
            <span
              class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
            >
              Stock Available
            </span>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-qwvqr0-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Add to Cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
